docs/standard/native-interop/tutorial-comwrappers.md

```diff
@@ -0,0 +1,430 @@
+
+```csharp
+protected override unsafe ComInterfaceEntry* ComputeVtables(object obj, CreateComInterfaceFlags flags, 
+out int count)
+{
+    if (obj is DemoImpl)
+    {
+        count = s_DemoImplDefinitionLen;
+        return s_DemoImplDefinition;
+    }
+
+    // Unknown type
+    count = 0;
+    return null;
+}
+```
+
+Once you've implemented the `ComputeVtables()` method, the `ComWrappers` subclass will be able to produce Managed Object Wrappers for instances of `DemoImpl`. Be aware that the returned Managed Object Wrapper from the call to [`GetOrCreateComInterfaceForObject()`](xref:System.Runtime.InteropServices.ComWrappers.GetOrCreateComInterfaceForObject%2A) is of type `IUnknown*`. If the native API that's being passed to the wrapper requires a different interface, a [`Marshal.QueryInterface()`][api_marshalqueryinterface] for that interface must be performed.
+
+```csharp
+var cw = new DemoComWrappers();
+var demo = new DemoImpl();
+IntPtr ccw = cw.GetOrCreateComInterfaceForObject(demo, CreateComInterfaceFlags.None);
+```
+
+### Step 3 &ndash; Implement `CreateObject()`
+
+Constructing a Native Object Wrapper has more implementation options and a great deal more nuance than constructing a Managed Object Wrapper. The first question to address is how permissive the `ComWrappers` subclass will be in supporting COM types. To support all COM types, which is possible, you'll need to write a substantial amount of code or employ some clever uses of `Reflection.Emit`. For this tutorial, you'll only support COM instances that implement both `IDemoGetType` and `IDemoStoreType`. Since you know there is a finite set and have restricted that any supplied COM instance must implement both interfaces, you could provide a single, statically defined wrapper; however, dynamic cases are common enough in COM that we'll explore both options.
+
+#### Static Native Object Wrapper
+
+Let's look at the static implementation first. The static Native Object Wrapper involves defining a managed type that implements the .NET interfaces and can forward the calls on the managed type to the COM instance. A rough outline of the static wrapper follows.
+
+```csharp
+class DemoNativeStaticWrapper
+    : IDemoGetType
+    , IDemoStoreType
+{
+    public string? GetString() =>
+        throw new NotImplementedException();
+
+    public void StoreString(int len, string? str) =>
+        throw new NotImplementedException();
+}
+```
+
+To construct an instance of this class and provide it as a wrapper, you must define some policy. If this type is used as a wrapper, it would seem that since it implements both interfaces, the underlying COM instance should implement both interfaces too. Given that you're adopting this policy, you'll need to confirm this through calls to [`Marshal.QueryInterface()`][api_marshalqueryinterface] on the COM instance.
+
+```csharp
+int hr = Marshal.QueryInterface(ptr, ref IDemoGetType.IID_IDemoGetType, out IntPtr IDemoGetTypeInst);
```

Comment:
I guess a C# rant - the `IID_IDemoGetType` can't be a readonly field - since the "ref" here is "in/out".

Reply:
I know, I know...